### ISSUE TITLE
YAML loading is terribly unsafe

### ISSUE THREAD

**Issue by mrdaemon** [ccfreak2k/BESM-Character-Resolver#3]
Reading serialized python classes in this case isn't the most efficient way to go at it. It makes the resulting file unreadable, is unsafe (because one could serialize executable code and have your program run it) and is also slightly clumsy should you modify the core set of properties.

I will hack at this issue to make it more efficient and a lot safer.

**Comment by Knifa** [comment]
Yeah I realise why this is utterly terrible now. I thought of a better solution but I'm not sure if I'm thinking about it entirely wrong. I realise there's safe_load() or something as well, apparently?

I assume modifying the classes we need to save to use PyYAML properly is the way to go right? Using the constructor and representer business, yes?

**Comment by mrdaemon** [comment]
Yes I think that would be the best -- we could just make a simple document containing values, and in the constructor (or the load method) parse them into a dict, and map the keys to the class attributes/properties.

Most yaml types translate directly into python collections :) So unless there is a direct need to actually serialize a python object, I'd stick to plain yaml.

**Comment by Knifa** [comment]
I did a bit of work on it in the Character module that can maybe be branched out into a full safe loader. I'm not sure if it's dumb or not. Basically I was too lazy to type out all the arguments it would need manually because there's a lot of them.

It might be an idea to pass a "safe list" of attributes to save and load instead.

**Comment by Knifa** [comment]
I've wrote a solution which uses an optional whitelist of attributes to save seems pretty good but it probably needs tidied up a bit with added error checking and so on. Check [yamlSafeLoader.py](https://github.com/ccfreak2k/BESM-Character-Resolver/blob/master/besmcr/yamlSafeLoader.py).

**EDIT:** Just realised this entirely doesn't work if there are non-builtin Python objects inside the serialised class, as the inner object isn't parsed by our own thing.

**EDIT 2:** Though, actually the way it's saved shouldn't be like this anyway.

**EDIT 3:** It might be an idea to use PyYAMLs built in stuff for adding tags still because it'll still make it tidier I guess.

**EDIT 4:** Regardless this might still be useful for compatibility because it can load files from other versions that are lacking the attributes.
